File: algorithms/169.Majority_Element/Majority_Element_Moore_n.py
class Solution:
    # @param num, a list of integers
    # @return an integer
    def majorityElement(self, num):
        cnt = 0
        for i in num:
            if cnt == 0:
                ret = i
            cnt += 1 if ret == i else -1
        return ret
class Solution:

    def majorityNumber(self, nums):
        candidate = None
        count = 0
        for num in nums:
            if count == 0:
                candidate = num
                count += 1
            elif candidate == num:
                count += 1
            else:
                count -= 1
        return candidate
# debug
s = Solution()
print s.majorityElement([1, 2, 2, 2, 3])

File: algorithms/169.Majority_Element/Majority_Element_Moore_n.java
public class Solution {
    /**
     * @param nums: a list of integers
     * @return: find a  majority number
     */
    public int majorityNumber(ArrayList<Integer> nums) {
        int count = 0, candidate = -1;
        for (int i = 0; i < nums.size(); i++) {
            if (count == 0) {
                candidate = nums.get(i);
                count = 1;
            } else if (candidate == nums.get(i)) {
                count++;
            } else {
                count--;
            }
        }
        return candidate;
    }
}